modules/gapi/src/streaming/gstreamer/gstreamer_pipeline_facade.cpp

```diff
@@ -0,0 +1,329 @@
+
+    if (!m_isPrerolled.load(std::memory_order_acquire))
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+
+        if(!m_isPrerolled.load(std::memory_order_relaxed))
+        {
+            PipelineState state = queryState();
+
+            // Only move forward in the pipeline's state machine
+            GAPI_Assert(state.current != GST_STATE_PLAYING);
+
+            GAPI_Assert(state.pending == GST_STATE_VOID_PENDING);
+            GstStateChangeReturn status = gst_element_set_state(m_pipeline, GST_STATE_PAUSED);
+            checkBusMessages();
+            if (status == GST_STATE_CHANGE_NO_PREROLL)
+            {
+                status = gst_element_set_state(m_pipeline, GST_STATE_PLAYING);
+                m_isPlaying.store(true);
+            }
+            verifyStateChange(status);
+
+            m_isPrerolled.store(true, std::memory_order_release);
+        }
+    }
+}
+
+void GStreamerPipelineFacade::play()
+{
+    if (!m_isPlaying.load(std::memory_order_acquire))
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+
+        if (!m_isPlaying.load(std::memory_order_relaxed))
+        {
+            setState(GST_STATE_PLAYING);
+            m_isPlaying.store(true, std::memory_order_release);
+            m_isPrerolled.store(true);
+        }
+    }
+}
+
+bool GStreamerPipelineFacade::isPlaying() {
+    return m_isPlaying.load();
+}
+
+std::vector<GstElement*> GStreamerPipelineFacade::getElements(
+    std::function<bool(GstElement*)> comparator)
+{
+    std::vector<GstElement*> outElements;
+    GStreamerPtr<GstIterator> it(gst_bin_iterate_elements(GST_BIN(m_pipeline.get())) /* transfer full ???*/);
+
+    gboolean done = false;
+    GValue value = G_VALUE_INIT;
+    while (!done)
+    {
+        switch (gst_iterator_next(it, &value))
+        {
+            case GST_ITERATOR_OK:
+            {
+                GstElement* element = GST_ELEMENT(g_value_get_object(&value)); /* transfer none */
+                if (comparator(element))
+                {
+                    outElements.push_back(GST_ELEMENT(element));
+                }
+
+                g_value_unset(&value);
+                break;
+            }
+            case GST_ITERATOR_RESYNC:
+            {
+                gst_iterator_resync(it);
+                break;
+            }
+            case GST_ITERATOR_ERROR:
+            case GST_ITERATOR_DONE:
+            {
+                done = TRUE;
+                break;
+            }
+        }
+    }
+
+    return outElements;
+}
+
+PipelineState GStreamerPipelineFacade::queryState()
+{
+    GAPI_Assert(m_pipeline && GST_IS_ELEMENT(m_pipeline.get()) &&
+                "GStreamer pipeline has not been created!");
+
+    PipelineState state;
+    GstClockTime timeout = 5 * GST_SECOND;
+    gst_element_get_state(m_pipeline, &state.current, &state.pending, timeout);
+
+    return state;
+}
+
+void GStreamerPipelineFacade::setState(GstState newState)
```

Comment:
IMHO, in my practice such kind of functions might return old value - it may be useful sometimes (comment not for address)